chia/full_node/fee_tracker.py

```diff
@@ -0,0 +1,577 @@
+
+    # For each bucket xL
+    # Count the total number of txs in each bucket
+    # Track historical moving average of this total over block
+    tx_ct_avg: List[float]
+
+    # Count the total number of txs confirmed within Y blocks in each bucket
+    # Track the historical moving average of these totals over blocks
+    confirmed_average: List[List[float]]  # confirmed_average [y][x]
+
+    # Track moving average of txs which have been evicted from the mempool
+    # after failing to be confirmed within Y block
+    failed_average: List[List[float]]  # failed_average [y][x]
+
+    # Sum the total fee_rate of all txs in each bucket
+    # Track historical moving average of this total over blocks
+    m_fee_rate_avg: List[float]
+
+    decay: float
+
+    # Resolution of blocks with which confirmations are tracked
+    scale: int
+
+    # Mempool counts of outstanding transactions
+    # For each bucket x, track the number of transactions in mempool
+    # that are unconfirmed for each possible confirmation value y
+    unconfirmed_txs: List[List[int]]
+    # transactions still unconfirmed after get_max_confirmed for each bucket
+    old_unconfirmed_txs: List[int]
+    max_confirms: int
+    fee_store: FeeStore
+
+    def __init__(
+        self,
+        buckets: List[float],
+        sorted_buckets: SortedDict,
+        max_periods: int,
+        decay: float,
+        scale: int,
+        log: logging.Logger,
+        fee_store: FeeStore,
+        my_type: str,
+    ):
+        self.buckets = buckets
+        self.sorted_buckets = sorted_buckets
+        self.confirmed_average = [[] for _ in range(0, max_periods)]
+        self.failed_average = [[] for _ in range(0, max_periods)]
+        self.decay = decay
+        self.scale = scale
+        self.max_confirms = self.scale * len(self.confirmed_average)
+        self.log = log
+        self.fee_store = fee_store
+        self.type = my_type
+        self.max_periods = max_periods
+
+        for i in range(0, max_periods):
+            self.confirmed_average[i] = [0 for _ in range(0, len(buckets))]
+            self.failed_average[i] = [0 for _ in range(0, len(buckets))]
+
+        self.tx_ct_avg = [0 for _ in range(0, len(buckets))]
+        self.m_fee_rate_avg = [0 for _ in range(0, len(buckets))]
+
+        self.unconfirmed_txs = [[] for _ in range(0, self.max_confirms)]
+        for i in range(0, self.max_confirms):
+            self.unconfirmed_txs[i] = [0 for _ in range(0, len(buckets))]
+
+        self.old_unconfirmed_txs = [0 for _ in range(0, len(buckets))]
+
+    def get_bucket_index(self, fee_rate: float) -> int:
+        if fee_rate in self.sorted_buckets:
+            bucket_index = self.sorted_buckets[fee_rate]
+        else:
+            bucket_index = self.sorted_buckets.bisect_left(fee_rate) - 1
+
+        return int(bucket_index)
+
+    def tx_confirmed(self, blocks_to_confirm: int, item: MempoolItem) -> None:
+        if blocks_to_confirm < 1:
```

Comment:
shouldn't the minimal blocks to confirm be 1 ?

maybe im misunderstanding but seems like an error message is needed

Reply:
Yes, it should never be called without blocks to confirm!